src/ctapipe/monitoring/calculator.py

```diff
@@ -0,0 +1,327 @@
+        if stats_aggregator is None:
+            for _, _, name in self.stats_aggregator_type:
+                self.stats_aggregator[name] = StatisticsAggregator.from_name(
+                    name, subarray=self.subarray, parent=self
+                )
+        else:
+            name = stats_aggregator.__class__.__name__
+            self.stats_aggregator_type = [("type", "*", name)]
+            self.stats_aggregator[name] = stats_aggregator
+
+        # Initialize the instances of OutlierDetector
+        self.outlier_detectors = {}
+        if self.outlier_detector_list is not None:
+            for outlier_detector in self.outlier_detector_list:
+                self.outlier_detectors[
+                    outlier_detector["apply_to"]
+                ] = OutlierDetector.from_name(
+                    name=outlier_detector["name"],
+                    validity_range=outlier_detector["validity_range"],
+                    subarray=self.subarray,
+                    parent=self,
+                )
+
+    def first_pass(
+        self,
+        table,
+        tel_id,
+        masked_pixels_of_sample=None,
+        col_name="image",
+    ) -> Table:
+        """
+        Calculate the monitoring data for a given set of events with non-overlapping aggregation chunks.
+
+        This method performs the first pass over the provided data table to calculate
+        various statistics for calibration purposes. The statistics are aggregated with
+        non-overlapping chunks (``chunk_shift`` set to None), and faulty pixels are detected
+        using a list of outlier detectors.
+
+
+        Parameters
+        ----------
+        table : astropy.table.Table
+            DL1-like table with images of shape (n_images, n_channels, n_pix), event IDs and
+            timestamps of shape (n_images, )
+        tel_id : int
+            Telescope ID for which the calibration is being performed
+        masked_pixels_of_sample : ndarray, optional
+            Boolean array of masked pixels of shape (n_pix, ) that are not available for processing
+        col_name : str
+            Column name in the table from which the statistics will be aggregated
+
+        Returns
+        -------
+        astropy.table.Table
+            Table containing the aggregated statistics, their outlier masks, and the validity of the chunks
+        """
+        # Get the aggregator
+        aggregator = self.stats_aggregator[self.stats_aggregator_type.tel[tel_id]]
+        # Pass through the whole provided dl1 table
+        aggregated_stats = aggregator(
+            table=table,
+            masked_pixels_of_sample=masked_pixels_of_sample,
+            col_name=col_name,
+            chunk_shift=None,
+        )
+        # Detect faulty pixels with multiple instances of ``OutlierDetector``
+        outlier_mask = np.zeros_like(aggregated_stats["mean"], dtype=bool)
+        for aggregated_val, outlier_detector in self.outlier_detectors.items():
```

Comment:
Maybe we should store a bitmask / boolean array of which of the detectors rejected the sample as outlier?

Like this, you cannot really understand why an event was rejected.

So maybe, compute all masks for all detectors in an nd array, then apply `or` in the end.

Reply:
We do not reject here on event-level but only on pixel-level. Currently we have for each calibration type (_pedestal_, _flatfield_, _timecalibration_, _pedestalvariance_) an outlier mask with shape of `(n_channels, n_pixels)`. Storing per detectors for each calibration type seems a little bit heavy and would make the downstream code a little bit more complex. I would only add this if that is really needed for the monitoring of the data quality.

Reply:
I am not saying that you should *use* that mask for the analysis. Just for understanding *why* a pixel was deemed unusable. Which is very important information e.g. to distinguish a hardware error from just reduced HV due to a star.

Reply:
Okay, makes sense! I just pushed a commit. Does it look good to you?